### PR TITLE
feat(loki-canary): Add support to push logs directly to Loki.

### PR DESCRIPTION
**Pull request by kavirajk**
**What this PR does / why we need it**:
Add additional `push` mode to loki-canary, which pushes the logs directly to given Loki URL as it generates logs.

The real function of Loki Canary is to act like a tenant and help us know the whether Loki installation is working as perceived by a real tenant. 

Main rationale for this additional push mode is to make canary more standalone without needing for `promtail` (or `grafana-agent`) to scrape it's logs and send to loki, with this change, Loki canary happily tests Loki behavior without needing any other dependencies.

**NOTE**:
If you run Loki behind any proxy that has different authorization policies to READ or WRITE to Loki, then  important change that canary operator need to be aware of it, now the user credentials they pass via `-user` and `-pass` to access loki endpoints need to have both `read' and `write` permissions (previously canary just query the logs where as promtail is the one pushes the logs, so just READ permissions was sufficient).

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:
There will be follow up PR(s) to cleanup, particularly `reader` and `comparitor` component in terms of logging with proper logger. Rationale is to keep the changes small per PR to make it easy to review.

This changes were tested it in one of the internal Loki dev cell.

**This PR is a no-op if this new `push` flag is disabled (it's disabled by default)**

**Checklist**
- [x] Documentation added
- [x] Tests updated
- [x] Is this an important fix or new feature? Add an entry in the `CHANGELOG.md`.
- [x] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

**Review by MasslessParticle — APPROVED** [review]

File: cmd/loki-canary/main.go
Comment (dannykopping) [inline cmd/loki-canary/main.go:150]: Why not use the logger here?

File: cmd/loki-canary/main.go
Comment (dannykopping) [inline cmd/loki-canary/main.go:185]: Why was this added?

File: pkg/canary/writer/push.go
Comment (dannykopping) [inline pkg/canary/writer/push.go:37]:
```suggestion
// Push is a io.Writer, that writes given log entries by pushing
```

File: pkg/canary/writer/push.go
Comment (dannykopping) [inline pkg/canary/writer/push.go:39]: Why no batching?

File: pkg/canary/writer/push.go
Comment (dannykopping) [inline pkg/canary/writer/push.go:198]: We should make it very explicit that this push mode does not retry when rate-limited, or when receiving 5xx responses.

I'm skeptical about this, though, because it means that the canary may show that a Loki cell is unhealthy even when things are fine, because of a transient error or a rate-limit.

File: pkg/canary/writer/push_test.go
Comment (dannykopping) [inline pkg/canary/writer/push_test.go:34]: Nice test!

**Review by dannykopping — CHANGES_REQUESTED** [review]
Raised a couple points here, let's address before we merge because I see you already have one approval.

File: pkg/canary/writer/push.go
Comment (kavirajk) [reply pkg/canary/writer/push.go:39]: The main rationale is two folded

1. Handling batch means, we need to take care of the lifecycle of the pusher during shutdown, say flush the batch before stopping. Kinda complicates the existing canary model a bit.
2.Given we push 1 or 2 logs (with just tiny payload) per sec. Didn't wanted to add these complexity now.

File: cmd/loki-canary/main.go
Comment (kavirajk) [reply cmd/loki-canary/main.go:185]: Oops. `os.Interrupt` is same as `syscall.SIGINT`. Overlooked. will remove

File: cmd/loki-canary/main.go
Comment (kavirajk) [reply cmd/loki-canary/main.go:150]: Currently, before this PR, canary was never using proper logger and were writing directly to os.Stderr and os.Stdout (both in `main.go` and other components).

I just wanted to keep it consistent for this PR (focussed on only changes related to `push`) and do cleanup in separate PR as mentioned in the reviewer note
```
There will be follow up PR(s) to cleanup, particularly reader and comparitor component in terms of logging with proper logger. Rationale is to keep the changes small per PR to make it easy to review.
```

File: pkg/canary/writer/push.go
Comment (kavirajk) [reply pkg/canary/writer/push.go:198]: Yea right. there is no backoff or retry mechanism in this push mode currently. that was intentional for the same purpose you mentioned. It's just a normal http client with timeout. curious what made you to suspect this would retry?

or do you want to add some comment for clarification?

Comment (kavirajk) [reply pkg/canary/writer/push.go:198]: >I guess because any transient failure in the Loki cell which self-heals will look like a failure to the canary, and give the operator an incorrect impression

So I assume you want to have retry mechanism similar to promtail. sounds reasonable. will add it

Comment (dannykopping) [reply pkg/canary/writer/push.go:198]: Yeah maybe an exponential backoff with a configurable number of retries
Doesn't have to be too fancy

Comment (kavirajk) [reply pkg/canary/writer/push.go]: Done :+1:

File: pkg/canary/writer/push.go
Comment (dannykopping) [inline pkg/canary/writer/push.go:211]: Is this err logged anywhere?

**Review by dannykopping — COMMENTED** [review]
Thanks @kavirajk! One comment about logging, then I think we're good to go

File: pkg/canary/writer/push.go
Comment (kavirajk) [reply pkg/canary/writer/push.go:215]: It is returned from this func at the end. And will be logged by the client calling `.Write()`.
https://github.com/grafana/loki/blob/kavirajk%2Fsupport-push-mode-canary/pkg/canary/writer/writer.go#L100

Comment (dannykopping) [reply pkg/canary/writer/push.go:215]: Ah, gotcha. Thanks.

**Review by dannykopping — APPROVED** [review]
LGTM!

**Pull request merged**
